telemetryDeviceDumper: segmentation fault when trying to save data

After #108 I managed to run with the gazebo icub simulator and I get this segmentation fault when it tried to save the files:

```
[ERROR] yarprobotinterface intercepted a segmentation fault caused by a faulty plugin:
[ERROR] /lib/x86_64-linux-gnu/libc.so.6(+0x18e964) [0x7f2ccb5ab964]
Trace requested at /home/ngenesio/robotology/robotology-superbuild/robotology/YARP/src/yarprobotinterface/Module.cpp:90 by code called from:
/home/ngenesio/robotology/robotology-superbuild/build/install/lib/libYARP_os.so.3(_Z16yarp_print_traceP8_IO_FILEPKcj+0x3b) [0x7f2ccb8da39b]
yarprobotinterface(+0x76a6) [0x55648f13c6a6]
/lib/x86_64-linux-gnu/libc.so.6(+0x46210) [0x7f2ccb463210]
/lib/x86_64-linux-gnu/libc.so.6(+0x18e964) [0x7f2ccb5ab964]
/usr/lib/x86_64-linux-gnu/libmatio.so.9(Mat_VarCreate+0x4db) [0x7f2cb06cf09b]
/home/ngenesio/robotology/robotology-superbuild/build/install/lib/libmatioCpp.so.0.1.0(_ZN8matioCpp8Variable18initializeVariableERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKNS_12VariableTypeERKNS_9ValueTypeENS_4SpanIKmLln1EEEPv+0x167) [0x7f2cb
072d127]
/home/ngenesio/robotology/robotology-superbuild/build/install/lib/yarp/yarp_telemetryDeviceDumper.so(_ZN4yarp9telemetry13BufferManagerIdE10saveToFileEb+0x592) [0x7f2cb0783fc2]
/home/ngenesio/robotology/robotology-superbuild/build/install/lib/yarp/yarp_telemetryDeviceDumper.so(_ZN4yarp9telemetry13BufferManagerIdE12periodicSaveEv+0x1a2) [0x7f2cb07851a2]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xd6d84) [0x7f2ccb702d84]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x9609) [0x7f2ccb17c609]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x43) [0x7f2ccb53f293]
[DEBUG] yarprobotinterface running happily
Illegal instruction
```
This bug has been introduced by the recent changes( #103 #104 #105 #106)

It is under investigation

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -138,7 +138,7 @@ if(BUILD_TESTING)
     find_package(Catch2 REQUIRED)
   else()
     include(FetchContent)
-    
+
     message(STATUS "Fetching Catch2...")
 
     FetchContent_Declare(Catch2
@@ -152,3 +152,5 @@ if(BUILD_TESTING)
 endif()
 
 set_property(GLOBAL PROPERTY USE_FOLDERS 1)
+
+include(AddUninstallTarget)

--- a/src/libYARP_telemetry/src/yarp/telemetry/Buffer.h
+++ b/src/libYARP_telemetry/src/yarp/telemetry/Buffer.h
@@ -115,6 +115,16 @@ public:
     }
 
     /**
+     * @brief Get the capacity of Buffer
+     *
+     * @return size_t The capacity of the buffer.
+     */
+    size_t capacity() const {
+        return m_buffer_ptr->capacity();
+
+    }
+
+    /**
      * @brief Return true if the Buffer is empty, false otherwise.
 
      *
@@ -130,6 +140,15 @@ public:
      */
     void resize(size_t new_size) {
         return m_buffer_ptr->resize(new_size);
+    }
+
+    /**
+     * @brief Change the capacity of the Buffer.
+     *
+     * @param[in] new_size The new size.
+     */
+    void set_capacity(size_t new_size) {
+        return m_buffer_ptr->set_capacity(new_size);
     }
 
     /**

--- a/src/libYARP_telemetry/src/yarp/telemetry/BufferManager.h
+++ b/src/libYARP_telemetry/src/yarp/telemetry/BufferManager.h
@@ -126,7 +126,7 @@ public:
      */
     bool configure(const BufferConfig& _bufferConfig) {
         bool ok{ true };
-        resize(_bufferConfig.n_samples);
+        set_capacity(_bufferConfig.n_samples);
         m_bufferConfig = _bufferConfig;
         if (!_bufferConfig.channels.empty()) {
             ok = ok && addChannels(_bufferConfig.channels);
@@ -185,11 +185,21 @@ public:
      * @param[in] new_size The new size to be resized to.
      */
     void resize(size_t new_size) {
-        if (new_size == m_bufferConfig.n_samples) {
-            return;
-        }
         for (auto& [var_name, buffInfo] : m_buffer_map) {
             buffInfo.m_buffer.resize(new_size);
+        }
+        m_bufferConfig.n_samples = new_size;
+        return;
+    }
+
+    /**
+     * @brief Set the capacity of Buffer/s.
+     *
+     * @param[in] new_size The new size.
+     */
+    void set_capacity(size_t new_size) {
+        for (auto& [var_name, buffInfo] : m_buffer_map) {
+            buffInfo.m_buffer.set_capacity(new_size);
         }
         m_bufferConfig.n_samples = new_size;
         return;
@@ -385,7 +395,7 @@ private:
     void periodicSave()
     {
         std::unique_lock<std::mutex> lk_cv(m_mutex_cv);
-        
+
         auto timeout =  std::chrono::duration<double>(m_bufferConfig.save_period);
         // For avoiding spurious wake up, the lambda check that the threads wake up only if we are trying to close
         // (additionally to the timeout expiration)

--- a/test/BufferManagerTest.cpp
+++ b/test/BufferManagerTest.cpp
@@ -234,7 +234,7 @@ TEST_CASE("Buffer Manager Test")
         REQUIRE(bm.addChannel(var_two));
 
         bufferConfig.description_list = { "Be", "Or not to be" };
-        bufferConfig.filename = "buffer_manager_test_resize";
+        bufferConfig.filename = "buffer_manager_test_set_capacity";
         bufferConfig.n_samples = 20;
         bufferConfig.data_threshold = 10;
         bufferConfig.save_periodically = false;

--- a/test/BufferManagerTest.cpp
+++ b/test/BufferManagerTest.cpp
@@ -176,12 +176,13 @@ TEST_CASE("Buffer Manager Test")
 
         bufferConfig.description_list = { "Be", "Or not to be" };
         bufferConfig.filename = "buffer_manager_test_resize";
-        bufferConfig.n_samples = 20;
         bufferConfig.data_threshold = 10;
         bufferConfig.save_periodically = false;
 
-        // This will resize the buffers
         REQUIRE(bm.configure(bufferConfig));
+
+        // Test resize
+        bm.resize(20);
 
         for (int i = 0; i < 40; i++) {
             bm.push_back({ i }, "one");
@@ -211,7 +212,7 @@ TEST_CASE("Buffer Manager Test")
         bufferConfig.auto_save = true;
         bufferConfig.save_period = 3600; // Save every hour.
 
-        // This will resize the buffers
+        // This will set_capacity the buffers
         REQUIRE(bm.configure(bufferConfig));
 
         for (int i = 0; i < 40; i++) {
@@ -219,6 +220,38 @@ TEST_CASE("Buffer Manager Test")
             yarp::os::Time::delay(0.01);
             bm.push_back({ i + 1 }, "two");
         }
+
+    }
+
+    SECTION("Test set_capacity") {
+        yarp::telemetry::BufferManager<int32_t> bm;
+        yarp::telemetry::BufferConfig bufferConfig;
+
+        yarp::telemetry::ChannelInfo var_one{ "one", {1,1} };
+        yarp::telemetry::ChannelInfo var_two{ "two", {1,1} };
+        // First add channels that will be handling empty buffers
+        REQUIRE(bm.addChannel(var_one));
+        REQUIRE(bm.addChannel(var_two));
+
+        bufferConfig.description_list = { "Be", "Or not to be" };
+        bufferConfig.filename = "buffer_manager_test_resize";
+        bufferConfig.n_samples = 20;
+        bufferConfig.data_threshold = 10;
+        bufferConfig.save_periodically = false;
+
+        // This will resize the buffers
+        REQUIRE(bm.configure(bufferConfig));
+
+        // Try to allocate a bigger buffer than what we need
+        bm.set_capacity(2000);
+
+        for (int i = 0; i < 40; i++) {
+            bm.push_back({ i }, "one");
+            yarp::os::Time::delay(0.01);
+            bm.push_back({ i + 1 }, "two");
+        }
+
+        REQUIRE(bm.saveToFile());
 
     }
 

